### PR TITLE
Refactor for abstraction

### LINKED ISSUE #3: Use of int causes overflows and arch-specific variability
Using the architecture-independent `int` type will cause 32-bit binaries to overflow if more than `21,474,836.47` is used in a base-100 currency, introducing architecture-dependent behaviour differences which may not be caught.

This can be resolved by switching amounts to use the `int64` data type. This will still work on 32-bit architectures, at the slight expense of performance, and allow base-100 currency values up to an amount of `92,233,720,368,547,758.07` ... More than likely enough for most consumers of this package :)

### PR DESCRIPTION
**Pull request by tylerb**
Fix #3
Address #6

- Amount struct with conversion methods and hidden internals
- Remove all pointers and dereferences. Unneeded. This should be faster (stay on stack)
- Remove calculator file and move all calculations to new Amount object

Due to the Amount struct hiding the internals, we can change the implementation to a `Big` implementation in the future, if necessary, without impacting existing code. We would simply add a `Big()` method to return the internal as a `Big`, as well as alter how the `toAmount` function parses the incoming values.

### CODE REVIEW

**Comment by tylerb** [comment]
@Rhymond please review and let me know what you think. I'd love to get this merged in ASAP as I intend to use it on a new project today.

**Comment by tylerb** [comment]
@Rhymond please note: the travis check is failing because it is misconfigured and cannot send the results to coveralls. The tests pass just fine.

**Comment by tylerb** [comment]
@bbrks I would value your input here as well.

**Comment by bbrks** [comment]
@tylerb My advice would be to define an interface containing all required methods (add, subtract, etc.) and then change all function signatures to accept any type fulfilling the interface.

Implement this interface with your current Amount methods, and it'll be trivial to add a Big implementation that can be used alongside this one.

E.g:

```go
type Money interface {
    Add(a Money) Money
    Sub(a Money) Money
    ...
    String() string
}
```

**Comment by tylerb** [comment]
@bbrks would the goal then be to have two separate implementations? One with `int64` and one with `Big`? That would allow the user to choose the one that is right for them, definitely.

I'm effectively doing what you described already, just with a struct instead of an interface, so it's a simple change.

File: money.go [hunk 1]
```diff
@@ -2,58 +2,77 @@ package money
...
+	case int8:
+		return Amount{value: int64(amt)}
+	case int16:
+		return Amount{value: int64(amt)}
+	case int32:
+		return Amount{value: int64(amt)}
+	case int64:
+		return Amount{value: amt}
+	}
+	panic(fmt.Sprintf("Unable to convert to Amount. Unsupported type: %s", reflect.TypeOf(amount).Name()))
```
Comment (Rhymond) [inline money.go]: Please don't use `panic` and leave error handling for users.

**Comment by tylerb** [comment]
@Rhymond I just pushed a huge refactor. Could you please take a look? cc @bbrks

**Comment by Rhymond** [comment]
@tylerb I didn't get any feedback on my answer.
Could you please update your branch from master? I updated travis file, builds should pass now.

**Comment by tylerb** [comment]
@Rhymond the push I just did made some abstractions that may affect your response.

I'm happy to move back to the calculator approach, but I find it unnecessary given that the Amount object now handles that work, and is abstracted as an interface so that the `Big` version of it can be created as well.

It's more a matter of style at this point, I think. Either way, two implementations have to be created.

Thoughts?

File: money.go (hunk 1 above)
Comment (tylerb) [inline money.go]: Done

**Comment by tylerb** [comment]
@Rhymond let me know when you've had a chance to look at the changes I've done. Happy to discuss your thoughts on it. It's your package after all 😄

**Comment by Rhymond** [comment]
Hi @tylerb,

I made a graph to illustrate, in what logic this library was design for.

untitled diagram

Orange parts as future plans. 
> It's more a matter of style at this point, I think. Either way, two implementations have to be created.

It's matter of structure and library extensibility

Int64
-
As we talk with @bbrks we should start using int64 type instead of int.
The idea is to assign `Amount` before creating `Money` structure when calling:
```Go
money.New(Amount, Currency)
```
use this for assigning Amount
```Go
function newAmount(amount interface{}) {
    switch amt := amount.(type) {
    case uint, uint8, uint16, uint32, uint64, int, int8, int16, int32:
        return Amount{value: int64(amt)}, nil
    case int64:
        return Amount{value: amt}, nil
    }

    return Amount{}, errors.New("Unsupported Amount type")
}

function New(Amount interface{}. Code string) (Money, error) {
    amt, err := amount.newAmount(Amount)

    if err != nil {
        return Money{}, err
    }

    return Money{
        amount:  amt,
        currency: currency.NewCurrency(Code),
    }, nil
}
```
Value field type of Amount structure should be changed to int64
```Go
type Amount struct {
    value int
}
```
to:
```Go
type Amount struct {
    value int64
}
```

Couple of observations
-
* There is no need to put types and interfaces in separate files
* All these Amount functions are redundant
```Go
Uint() uint
Uint8() uint8
Uint16() uint16
Uint32() uint32
Uint64() uint64
Int() int
Int8() int8
Int16() int16
Int32() int32
Int64() int64
```


There is definitely more to discuss because you made too much changes for one PR.
To be more precise, please address separate issues in separate PR's.

**Comment by tylerb** [comment]
I love the diagram and roadmap you've got here. It seems you've got it very well thought out.

I definitely went down the refactoring rabbit hole here. A bit overenthusiastic 😄 

I'll close this PR and do a small one to just get int64 functionality in there.

**Pull request closed without merging**